server/aws-lsp-identity/src/iam/utils.ts

```diff
@@ -7,27 +7,62 @@ import {
     GetMfaCodeResult,
     IamCredentials,
     Profile,
+    StsCredentialChangedParams,
 } from '@aws/language-server-runtimes/server-interface'
 import { AwsError, Observability } from '@aws/lsp-core'
 import { StsCache } from '../sts/cache/stsCache'
 import { StsAutoRefresher } from '../sts/stsAutoRefresher'
 import { ProfileStore } from '../language-server/profiles/profileService'
 
-// Simulate permissions on the identity associated with the credentials
-export async function simulatePermissions(
+export async function validatePermissions(
+    credentials: IamCredentials,
+    permissions: string[],
+    region?: string
+): Promise<boolean> {
+    const response = await simulatePermissions(credentials, permissions, region)
+    // If evaluation results are missing, assume caller does not have sufficient permissions
+    if (!response.EvaluationResults) {
+        return false
+    }
+    return response.EvaluationResults.every(result => result.EvalDecision === 'allowed')
+}
+
+export async function checkMfaRequired(
     credentials: IamCredentials,
     permissions: string[],
     region?: string
+): Promise<boolean> {
+    const response = await simulatePermissions(credentials, permissions, region)
+    // If evaluation results are missing, assume caller does not need MFA
+    if (!response.EvaluationResults) {
+        return false
+    }
+    return response.EvaluationResults?.some(result =>
+        result?.MissingContextValues?.includes('aws:MultiFactorAuthPresent')
+    )
+}
+
+export function throwOnInvalidCredentialId(iamCredentialId?: string): asserts iamCredentialId is string {
+    if (typeof iamCredentialId?.trim !== 'function' || !iamCredentialId?.trim()) {
+        throw new AwsError('IAM credential id is invalid.', AwsErrorCodes.E_INVALID_STS_CREDENTIAL)
+    }
+}
+
+// Simulate permissions on the identity associated with the credentials
+async function simulatePermissions(
+    credentials: IamCredentials,
+    permissions: string[],
+    region: string = 'us-east-1'
 ): Promise<SimulatePrincipalPolicyCommandOutput> {
     // Convert the credentials into an identity
-    const stsClient = new STSClient({ region: region || 'us-east-1', credentials: credentials })
+    const stsClient = new STSClient({ region: region, credentials: credentials })
```

Comment:
Why not keep the `region || defaultRegion` in place?  Did tsc complain since `region: string` is no longer optional?